pkg/scale/decode.go

```diff
@@ -0,0 +1,539 @@
+	for i := 0; i < temp.Elem().Len(); i++ {
+		elem := temp.Elem().Index(i)
+		err = ds.unmarshal(elem)
+		if err != nil {
+			return
+		}
+	}
+	dstv.Set(temp.Elem())
+	return
+}
+
+// decodeStruct comment about this
+func (ds *decodeState) decodeStruct(dstv reflect.Value) (err error) {
+	in := dstv.Interface()
+	_, indices, err := cache.fieldScaleIndices(in)
+	if err != nil {
+		return
+	}
+	temp := reflect.New(reflect.ValueOf(in).Type())
+	for _, i := range indices {
+		field := temp.Elem().Field(i.fieldIndex)
+		if !field.CanInterface() {
+			continue
+		}
+		err = ds.unmarshal(field)
+		if err != nil {
+			err = fmt.Errorf("%s, field: %+v", err, field)
+			return
+		}
+	}
+	dstv.Set(temp.Elem())
+	return
+}
+
+// decodeBool accepts a byte array representing a SCALE encoded bool and performs SCALE decoding
+// of the bool then returns it. if invalid returns an error
+func (ds *decodeState) decodeBool(dstv reflect.Value) (err error) {
+	rb, err := ds.ReadByte()
+	if err != nil {
+		return
+	}
+
+	var b bool
+	switch rb {
+	case 0x00:
+	case 0x01:
+		b = true
+	default:
+		err = fmt.Errorf("could not decode invalid bool")
+	}
+	dstv.Set(reflect.ValueOf(b))
+	return
+}
+
+// decodeUint will decode unsigned integer
+func (ds *decodeState) decodeUint(dstv reflect.Value) (err error) {
+	b, err := ds.ReadByte()
+	if err != nil {
+		return
+	}
+
+	in := dstv.Interface()
+	temp := reflect.New(reflect.TypeOf(in))
+	// check mode of encoding, stored at 2 least significant bits
+	mode := b & 3
+	switch {
+	case mode <= 2:
+		var val int64
+		val, err = ds.decodeSmallInt(b, mode)
+		if err != nil {
+			return
+		}
+		temp.Elem().Set(reflect.ValueOf(val).Convert(reflect.TypeOf(in)))
+		dstv.Set(temp.Elem())
+	default:
+		// >4 byte mode
+		topSixBits := b >> 2
+		byteLen := uint(topSixBits) + 4
+
+		buf := make([]byte, byteLen)
+		_, err = ds.Read(buf)
+		if err != nil {
+			return
+		}
+
+		var o uint64
+		if byteLen == 4 {
+			o = uint64(binary.LittleEndian.Uint32(buf))
+		} else if byteLen > 4 && byteLen <= 8 {
+			tmp := make([]byte, 8)
+			copy(tmp, buf)
+			o = binary.LittleEndian.Uint64(tmp)
+		} else {
+			err = errors.New("could not decode invalid integer")
+			return
+		}
+		dstv.Set(reflect.ValueOf(o).Convert(reflect.TypeOf(in)))
+	}
+	return
+}
+
+// decodeLength is helper method which calls decodeUint and casts to int
+func (ds *decodeState) decodeLength() (l int, err error) {
+	dstv := reflect.New(reflect.TypeOf(l))
+	err = ds.decodeUint(dstv.Elem())
+	if err != nil {
+		return
+	}
+	l = dstv.Elem().Interface().(int)
+	return
+}
+
+// decodeBytes is used to decode with a destination of []byte or string type
+func (ds *decodeState) decodeBytes(dstv reflect.Value) (err error) {
+	length, err := ds.decodeLength()
+	if err != nil {
+		return
+	}
+
+	b := make([]byte, length)
+	_, err = ds.Read(b)
+	if err != nil {
+		return
+	}
+
+	in := dstv.Interface()
+	inType := reflect.TypeOf(in)
+	dstv.Set(reflect.ValueOf(b).Convert(inType))
```

Comment:
would this panic if `b` cannot be converted to `inType`?

Reply:
yes the dst is always `[]byte`.  The big type switch is in `decodeState.unmarshal`